networkx/algorithms/tests/test_planarity.py

```diff
@@ -277,6 +277,20 @@ def test_counterexample_planar_recursive(self):
             G.add_node(1)
             get_counterexample_recursive(G)
 
+    def test_edge_removal_from_planar_embedding(self):
+        # PlanarEmbedding.check_structure() must succeed after edge removal
+        edges = ((0, 1), (1, 2), (2, 3), (3, 4), (4, 0), (0, 2), (0, 3))
+        G = nx.Graph(edges)
+        cert, P = nx.check_planarity(G)
+        assert cert is True
+        P.remove_edge(0, 2)
+        self.check_graph(P, is_planar=True)
+        P.add_half_edge_ccw(1, 3, 2)
+        P.add_half_edge_cw(3, 1, 2)
+        self.check_graph(P, is_planar=True)
+        P.remove_edges_from(((0, 3), (1, 3)))
+        self.check_graph(P, is_planar=True)
+
```

Comment:
Let's add tests for the forbidden functions. Maybe just something like this:
```suggestion

    def test_forbidden_methods(self):
        G = nx.Graph([(0, 1), (1, 2), (3, 4), (4, 5)])
        G.add_node(42)  # no exception
        G.add_nodes_from([(23, 24])  # no exception
        with pytest.raises(nx.NetworkXError):
            G.add_edge(1, 3)
        with pytest.raises(nx.NetworkXError):
            G.add_edges_from([(0, 2), (1, 4)])
        with pytest.raises(nx.NetworkXError):
            G.add_weighted_edges_from([(0, 2, 350), (1, 4, 125)])
```

Reply:
`test_planarity.py` is using the forbidden method `PlanarEmbedding.add_edge()` in several tests. This is exemplified by the following snippet:

```python
with pytest.raises(nx.NetworkXException):
       embedding = nx.PlanarEmbedding()
       embedding.add_edge(1, 2, ccw=2, cw=3)
       # continues with some operation that raises nx.NetworkXException
```

The test will pass, but for the wrong reason: `add_edge()` will raise the expected exception.

My first thought was to change all tests with this pattern to calling `nx.PlanarEmbedding(desired_graph_data)` instead of adding edges after the object is constructed. But this is not possible because there is no way of including node attributes in the constructor call (which would be required to set the `first_nbr` node attribute).

I thought of two alternatives to solve that problem (option B seems more elegant):

A) rewrite several tests in `test_planarity.py` to the structure:
```python
emb = nx.PlanarEmbedding(dict_of_dicts_with_edge_data)
nx.set_node_attributes(emb, dict_of_node_to_first_nbr_mapping, "first_nbr")
```

B) completely remove all references to the `"first_nbr"` node attribute from PlanarEmbedding, since it can be replaced by either:
- `bool(emb._succ[node])` when testing if `"first_nbr"` is defined
- `next(iter(emb._succ[node])` to get the first neighbor of `node`

I can write a PR for option B, which would need to be merged before this one. Please indicate which approach is preferred, @dschult

Reply:
I just made PR #7202 implementing option B described above.

Reply:
Thanks for adding PR #7202  And for working through all this. I prefer option B.
And I guess we should merge #7202 before this one then... right?

Reply:
I tried to stack this PR on top of #7202 by rebasing my fork's main branch to start from the del_first_nbr branch, but I think it didn't work as I expected.

I'm not sure what is the best approach regarding the git tree, but I'd say this PR is complete now.